crates/provider/src/lib.rs

```diff
@@ -2,85 +2,126 @@ mod errors;
+
+#[async_trait]
+pub trait ProviderNode {
+    async fn name(&self) -> String;
+
+    async fn endpoint(&self) -> Result<(IpAddr, Port), ProviderError>;
+
+    async fn mapped_port(&self, port: Port) -> Result<Port, ProviderError>;
+
+    async fn logs(&self) -> Result<String, ProviderError>;
+
+    async fn dump_logs(&self, dest: PathBuf) -> Result<(), ProviderError>;
+
+    async fn run_command(
         &self,
-        node: Self::Node,
-        // Files to inject, `before` we run the provider command.
-        files_inject: Vec<FileMap>,
-        // Files to get, `after` we run the provider command.
-        files_get: Vec<FileMap>,
-    ) -> Result<(), ProviderError>;
-    /// Copy a single file from node to local filesystem.
+        options: RunCommandOptions,
+    ) -> Result<ExecutionResult, ProviderError>;
+
+    async fn run_script(&self, options: RunScriptOptions)
+        -> Result<ExecutionResult, ProviderError>;
+
     async fn copy_file_from_node(
-        &mut self,
-        node_file_path: PathBuf,
-        local_file_path: PathBuf,
-    ) -> Result<(), ProviderError>;
-    /// Run a command inside the node.
-    async fn run_command(
         &self,
-        args: Vec<String>,
-        opts: NativeRunCommandOptions,
-    ) -> Result<RunCommandResponse, ProviderError>;
-    /// Run a script inside the node, should be a shell script.
-    /// zombienet will upload the content first.
-    async fn run_script(
-        &mut self,
-        identifier: String,
-        script_path: String,
-        args: Vec<String>,
-    ) -> Result<RunCommandResponse, ProviderError>;
-    async fn get_node_logs(&mut self, node_name: &str) -> Result<String, ProviderError>;
-    async fn dump_logs(&mut self, path: String, node_name: String) -> Result<(), ProviderError>;
-    async fn get_logs_command(&self, node_name: &str) -> Result<String, ProviderError>;
-    async fn pause(&self, node_name: &str) -> Result<(), ProviderError>;
-    async fn resume(&self, node_name: &str) -> Result<(), ProviderError>;
-    async fn restart(
-        &mut self,
-        node_name: &str,
-        after_sec: Option<u16>,
-    ) -> Result<bool, ProviderError>;
-    async fn get_node_info(&self, node_name: &str) -> Result<(IpAddr, Port), ProviderError>;
-    async fn get_node_ip(&self, node_name: &str) -> Result<IpAddr, ProviderError>;
-    async fn get_port_mapping(&self, port: Port, node_name: &str) -> Result<Port, ProviderError>;
-    async fn static_setup(&mut self) -> Result<(), ProviderError>;
-    async fn create_static_resource() -> Result<(), ProviderError> {
-        Ok(())
-    }
+        remote_src: PathBuf,
+        local_dest: PathBuf,
+    ) -> Result<(), ProviderError>;
 
-    // TODO(team): Do we need at this point to handle cleanner/pod-monitor?
+    async fn pause(&self) -> Result<(), ProviderError>;
+
+    async fn resume(&self) -> Result<(), ProviderError>;
+
+    async fn restart(&mut self, after: Option<Duration>) -> Result<(), ProviderError>;
```

Comment:
In a second thought, the only reason to this be a `mut ref` is because we want to store the `pid` of the process in the native provider. Maybe we can store elsewhere in the native provider and don't restrict the rest of them.

Reply:
Yes, it's a mistake, doesn't need to be mutable because we use an Arc here in ```inner``` and data inside it can be borrowed as mutable (we store the lock)